File: 42 Trapping Rain Water/42 Trapping Rain Water.ts
function trap(height: number[]): number {
  const leftMaxs: number[] = Array(height.length).fill(0);
  const rightMaxs: number[] = Array(height.length).fill(0);
  let max: number = 0;
  let result: number = 0;

  for (let i = 0; i < height.length; i++) {
    max = Math.max(max, height[i]);
    leftMaxs[i] = max;
  }

  max = 0;
  for (let i = height.length - 1; i >= 0; i--) {
    max = Math.max(max, height[i]);
    rightMaxs[i] = max;
  }

  for (let i = 0; i < height.length; i++) {
    result =
      result + Math.max(0, Math.min(leftMaxs[i], rightMaxs[i]) - height[i]);
  }

  return result;
}

File: 42 Trapping Rain Water/42 Trapping Rain Water.swift
class Solution {
  func trap(_ height: [Int]) -> Int {
    let n: Int = height.count
    var currentMax: Int = 0
    var leftMaxs: [Int] = Array(repeating: 0, count: n)
    var rightMaxs: [Int] = Array(repeating: 0, count: n)

    for i in 0..<height.count {
      currentMax = max(currentMax, height[i])
      leftMaxs[i] = currentMax
    }

    currentMax = 0

    for i in stride(from: height.count - 1, through: 0, by: -1) {
      currentMax = max(currentMax, height[i])
      rightMaxs[i] = currentMax
    }

    return height.enumerated().reduce(0, {accumulator, element in 
      accumulator + max(0, min(leftMaxs[element.0], rightMaxs[element.0]) - element.1)
    })
  }
}